### PR TITLE
Mandatory Update and Package Rollout Support

### LINKED ISSUE #2: Add support for Package Rollout and Mandatory Update
Need to review the new API's and see the best way to expose them.

### PR DESCRIPTION
**Pull request by HowardWolosky**
Allows users to specify that an application or flight submission
is a mandatory update, or to enable gradual rollout on the packages
for those submissions.

This also introduces new top-level commands to allow users
to further control package rollout (Get / Update / Complete / Stop).

Additionally, the `Format-ApplicationSubmission` and
`Format-ApplicationFlightSubmission` commands were updated to also print
out mandatory update and package rollout information.  No new Format
command was introduced for `Get-ApplicationSubmissionPackageRollout` or
`Get-ApplicationFlightSubmissionPackageRollout` since the returned content
was simple enough as-is.

Additionally:

  * Updated `Get-PiiSafeString` / `Get-SHA512Hash` to support receiving
    empty / null strings, as it's possible to call
    `Update-ApplicationSubmission` and `Update-ApplicationFlightSubmission`
    without providing a packagePath, and without this update that would cause
    a parameter binding failure in those scenarios

  * Updated all of the `Patch-*Submission` methods to first convert
    TargetDate to UTC before using it, as that is the expected behavior from
    the API.  The documentation and Comment-Based-Help has been updated to
    indicate to users that the time will be converted to UTC for them.

Resolves #2: Add support for Package Rollout and Mandatory Update

### CODE REVIEW

File: StoreBroker/StoreIngestionFlightingApi.ps1 [hunk 1]
```diff
@@ -1834,4 +1965,411 @@ function Start-ApplicationFlightSubmissionMonitor
...
+            [StoreBrokerTelemetryProperty]::FlightId = $FlightId
+            [StoreBrokerTelemetryProperty]::SubmissionId = $SubmissionId
+        }
+
+        $params = @{
+            "UriFragment" = "applications/$AppId/flights/$FlightId/submissions/$SubmissionId/packagerollout"
+            "Method" = "Get"
+            "Description" = "Getting package rollout on submission $SubmissionId for App: $AppId FlightId: $FlightId"
+            "AccessToken" = $AccessToken
+            "TelemetryEventName" = "GEt-ApplicationFlightSubmissionPackageRollout"
```
Comment (danbelcher-MSFT) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: typo

File: StoreBroker/StoreIngestionFlightingApi.ps1 [hunk 2]
```diff
@@ -1834,4 +1965,411 @@ function Start-ApplicationFlightSubmissionMonitor
...
+        throw
+    }
+}
+
+function Stop-ApplicationFlightSubmissionPackageRollout
+{
+<#
+    .SYNOPSIS
+        Stops the package rollout for the specified application submission.
+
```
Comment (danbelcher-MSFT) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: Do you care to explicitly note "application flight submission"?

File: StoreBroker/StoreIngestionFlightingApi.ps1 [hunk 3]
```diff
@@ -1834,4 +1965,411 @@ function Start-ApplicationFlightSubmissionMonitor
...
+        $telemetryProperties = @{
+            [StoreBrokerTelemetryProperty]::AppId = $AppId
+            [StoreBrokerTelemetryProperty]::FlightId = $FlightId
+            [StoreBrokerTelemetryProperty]::SubmissionId = $SubmissionId
+        }
+
+        $params = @{
+            "UriFragment" = "applications/$AppId/flights/$FlightId/submissions/$SubmissionId/finalizepackagerollout"
+            "Method" = "Post"
+            "Description" = "Finalzing package rollout on submission $SubmissionId for App: $AppId FlightId: $FlightId"
```
Comment (danbelcher-MSFT) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: same typo

File: StoreBroker/StoreIngestionFlightingApi.ps1 [hunk 4]
```diff
@@ -1834,4 +1965,411 @@ function Start-ApplicationFlightSubmissionMonitor
...
+
+        $null = Invoke-SBRestMethod @params
+
+        Write-Log "Package rollout for this submission has been finalized.  All users in this flight will now receive these packages."
+    }
+    catch [System.InvalidOperationException]
+    {
+        throw
+    }
 }
```
Comment (danbelcher-MSFT) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: newline at end of file

File: Documentation/USAGE.md [hunk 5]
```diff
@@ -228,8 +241,9 @@ fully override the publication and visibility of your app submission.
        Specify with `TargetPublishDate`
 
  * **`TargetPublishDate`** - The specific date/time that the submission should be published. To use
- this, you must specify `TargetPublishMode` as `SpecificDate`.  Using this value will override
- any value that might have been set by `-UpdatePublishModeAndVisibility`.
+   this, you must specify `TargetPublishMode` as `SpecificDate`.  Using this value will override
+   any value that might have been set by `-UpdatePublishModeAndVisibility`.  This value will be converted
+   to UTC automatically.
```
Comment (HowardWolosky) [inline Documentation/USAGE.md]: I've modified the code to only convert the date time if it's not already in UTC (by checking the .Kind property).  I've updated the documentation (in the .md and CBH) to be a bit more clear:
_This value will be converted to UTC automatically if it is not already in UTC._

File: Documentation/USAGE.md [hunk 6]
```diff
@@ -310,6 +344,31 @@ for your submission to enter into the Store certification process.
...
 > If using them for flight submissions, just be sure to also include the FlightId parameter.
 
+### Related Commands
+
+In addition to the commands above, there are some other commands that you may find useful along the way.
+
+#### Gradual Rollout Commands
+
+If you're not familiar with package rollout, it might be helpful to read more about it
+in the [Store documentatin](https://docs.microsoft.com/en-us/windows/uwp/publish/gradual-package-rollout).
```
Comment (HowardWolosky) [inline Documentation/USAGE.md]: fixed

File: Documentation/USAGE.md [hunk 7]
```diff
@@ -458,8 +529,29 @@ fully override the publication of your app's flight submission.
        `TargetPublishDate`
 
  * **`TargetPublishDate`** - The specific date/time that the submission should be published. To use
- this, you must specify `TargetPublishMode` as `SpecificDate`.  Using this value will override
- any value that might have been set by `-UpdatePublishMode`.
+   this, you must specify `TargetPublishMode` as `SpecificDate`.  Using this value will override
```
Comment (HowardWolosky) [inline Documentation/USAGE.md]: I think the intention there was to show that one was a switch (with no additional input) and the other wasn't.

File: StoreBroker/StoreBroker.psd1 [hunk 8]
```diff
@@ -6,11 +6,11 @@
...
     Copyright = 'Copyright (C) Microsoft Corporation.  All rights reserved.'
 
-    ModuleVersion = '1.2.2'
+    ModuleVersion = '1.3.0'
     Description = 'Provides command-line access to the Windows Store Submission REST API.'
 
     RootModule = 'StoreIngestionApi'
-
+;
     # Modules to import as nested modules of the module specified in RootModule/ModuleToProcess
```
Comment (HowardWolosky) [inline StoreBroker/StoreBroker.psd1]: yes, thanks

File: StoreBroker/StoreIngestionApplicationApi.ps1 [hunk 9]
```diff
@@ -1688,4 +1819,378 @@ function Complete-ApplicationSubmission
...
+            [StoreBrokerTelemetryProperty]::AppId = $AppId
+            [StoreBrokerTelemetryProperty]::SubmissionId = $SubmissionId
+        }
+
+        $params = @{
+            "UriFragment" = "applications/$AppId/submissions/$SubmissionId/packagerollout"
+            "Method" = "Get"
+            "Description" = "Getting package rollout on submission $SubmissionId for App: $AppId"
+            "AccessToken" = $AccessToken
+            "TelemetryEventName" = "GEt-ApplicationSubmissionPackageRollout"
```
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: Assuming you meant the `GEt`, fixed.

File: StoreBroker/StoreIngestionApplicationApi.ps1 [hunk 10]
```diff
@@ -1688,4 +1819,378 @@ function Complete-ApplicationSubmission
...
+        Write-Log "Package rollout for this submission has been updated to $Percentage%."
+
+        if ($Percentage -eq 100)
+        {
+            $output = @()
+            $output += "Changing the rollout percentage to 100% does not ensure that all of your customers will get the"
+            $output += "packages from the latest submissions, because some customers may be on OS versions that don't"
+            $output += "support rollout. You must finalize the rollout in order to stop distributing the older packages"
+            $output += "and update all existing customers to the newer ones by calling"
+            $output += "Complete-ApplicationFlightSubmissionPackageRollout -AppId $AppId -FlightId $FlightId -SubmissionId $SubmissionId"
```
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: done

File: StoreBroker/StoreIngestionApplicationApi.ps1 [hunk 11]
```diff
@@ -1688,4 +1819,378 @@ function Complete-ApplicationSubmission
...
+        If this switch is specified, long-running commands will run on the main thread
+        with no commandline status update.  When not specified, those commands run in
+        the background, enabling the command prompt to provide status information.
+
+    .EXAMPLE
+        Stop-ApplicationSubmissionPackageRollout 0ABCDEF12345 1234567890123456789
+
+        Halts the package rollout for submission 1234567890123456789 of application
+        0ABCDEF12345.  Users will now begin receiving the packages from the fallback
+        submission. The console window will show progress while awaiting for the respons
```
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: I'm not seeing the typo here.

File: StoreBroker/StoreIngestionApplicationApi.ps1 [hunk 12]
```diff
@@ -1688,4 +1819,378 @@ function Complete-ApplicationSubmission
...
+    {
+        $telemetryProperties = @{
+            [StoreBrokerTelemetryProperty]::AppId = $AppId
+            [StoreBrokerTelemetryProperty]::SubmissionId = $SubmissionId
+        }
+
+        $params = @{
+            "UriFragment" = "applications/$AppId/submissions/$SubmissionId/finalizepackagerollout"
+            "Method" = "Post"
+            "Description" = "Finalzing package rollout on submission $SubmissionId for App: $AppId"
```
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: fixed `Finalzing`

File: StoreBroker/StoreIngestionApplicationApi.ps1 [hunk 13]
```diff
@@ -1688,4 +1819,378 @@ function Complete-ApplicationSubmission
...
+
+        $null = Invoke-SBRestMethod @params
+
+        Write-Log "Package rollout for this submission has been finalized.  All users will now receive these packages."
+    }
+    catch [System.InvalidOperationException]
+    {
+        throw
+    }
 }
```
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: fixed

File: StoreBroker/StoreIngestionFlightingApi.ps1 (hunk 1 above)
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: Fixed `GEt`

File: StoreBroker/StoreIngestionFlightingApi.ps1 (hunk 2 above)
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: Yes. Fixed

File: StoreBroker/StoreIngestionFlightingApi.ps1 (hunk 3 above)
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: Fixed `Finalzing`

File: StoreBroker/StoreIngestionFlightingApi.ps1 (hunk 4 above)
Comment (HowardWolosky) [inline StoreBroker/StoreIngestionFlightingApi.ps1]: Fixed

File: Documentation/USAGE.md [hunk 14]
```diff
@@ -458,8 +529,29 @@ fully override the publication of your app's flight submission.
...
+   This value will be ignored if `-IsMandatoryUpdate` is not also provided. This value will be converted to
+   UTC automatically.
+
+You can also leverage gradual rollouts to limit the percentage of users who will be given the packages within
+this submission.
+
+ * **`PackageRolloutPercentage`** - _[0 - 100]_.  If specified, the packages in this submission will only
+   be rolled out to this percentage of your users.  You can later update this percentage by calling
+   `Update-ApplicationFlightSubmissionPackageRollout`, and can complete the gradual rollout by either
+   finalizing it with `Complete-ApplicationFlightSubmissionPackageRollout` (which will make the packages available
```
Comment (danbelcher-MSFT) [inline Documentation/USAGE.md]: At the point I made that comment, I hadn't gotten far enough to see that you added a warning when executing the command.  I'm happy that that information is included somewhere so no need to add it here if you think it's overkill.

File: StoreBroker/StoreIngestionApplicationApi.ps1 [hunk 15]
```diff
@@ -1439,7 +1441,13 @@ function Patch-ApplicationSubmission
     {
         if ($IsMandatoryUpdate)
         {
-            $PatchedSubmission.packageDeliveryOptions.mandatoryUpdateEffectiveDate = $MandatoryUpdateEffectiveDate.ToUniversalTime().ToString('o')
+            $utcDateTime =  $MandatoryUpdateEffectiveDate
+            if ($utcDateTime.Kind -ne [System.DateTimeKind]::Utc)
+            {
+                $utcDateTime = $utcDateTime.ToUniversalTime()
```
Comment (danbelcher-MSFT) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: I don't think we need to check the kind before making this call.  In almost all cases we will need to call ToUniversalTime.  In the edge case where they pass us a datetime that is already UTC, calling ToUniversalTime shouldn't have any effect and should be safe to do.

Comment (HowardWolosky) [inline StoreBroker/StoreIngestionApplicationApi.ps1]: That's a good point.  Will revert back to what I had before.

**Pull request merged**